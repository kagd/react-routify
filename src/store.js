import page from 'page'
import { EventEmitter } from 'events'
import getCurrentPath from './getCurrentPath'

const CHANGE_EVENT = 'CHANGE_EVENT'

class Store extends EventEmitter {
  constructor() {
    super()
    this.ctx = { path: getCurrentPath() }

    page('*', (ctx, next) => {
      this.emitChange(ctx)
      next()
    })
  }

  getPath() {
    return this.ctx.path
  }

  emitChange(ctx) {
    this.ctx = { path: ctx.path }
    this.emit(CHANGE_EVENT, ctx)
  }

  addChangeListener(callback) {
    this.on(CHANGE_EVENT, callback)
  }

  removeChangeListener(callback) {
    this.removeListener(CHANGE_EVENT, callback)
  }
}

export default new Store()
