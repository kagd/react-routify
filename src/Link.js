import React, { Component } from 'react'
import store from './store'

class Link extends Component {
  constructor(props) {
    super(props)

    this.state = {
      active: store.getPath() === props.to
    }
  }

  _onChange(ctx) {
    this.setState({
      active: ctx.path === this.props.to
    })
  }

  componentDidMount() {
    store.addChangeListener(this._onChange.bind(this))
  }

  componentWillUnmount() {
    store.removeChangeListener(this._onChange.bind(this))
  }

  render() {
    const {to, activeClassName, ...rest} = this.props

    let props = {
      href: to,
      ...rest
    }

    if(activeClassName && this.state.active) {
      if(props.className) {
        props.className += ' ' + activeClassName
      } else {
        props.className = activeClassName
      }
    }

    return <a {...props} />
  }
}

export default Link
