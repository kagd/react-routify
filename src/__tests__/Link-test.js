jest.mock('../store')

import React from 'react'
import renderer from 'react-test-renderer'
import Link from '../Link'
import store from '../store'

test('Link should render', () => {
  const link = <Link to="/nowhere" className="some-class" id="something" />
  const component = renderer.create(link)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Link should apply activeClassName', () => {
  store.getPath.mockImplementation(() => {
    return '/nowhere'
  })

  const link = <Link to="/nowhere" activeClassName="active" id="something" />
  const component = renderer.create(link)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Link should merge className and activeClassName', () => {
  store.getPath.mockImplementation(() => {
    return '/nowhere'
  })

  const link = <Link to="/nowhere" className="foo" activeClassName="active" id="something" />
  const component = renderer.create(link)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Link should update active state onRouteChange', () => {
  const callbacks = []

  store.getPath.mockImplementation(() => {
    return '/foo'
  })

  store.addChangeListener.mockImplementation(callback => {
    callbacks.push(callback)
  })

  const ui = (
    <div>
      <Link to="/foo" activeClassName="active">foo</Link>
      <Link to="/bar" activeClassName="active">bar</Link>
    </div>
  )

  const changeRoute = path => {
    callbacks.forEach(callback => {
      callback({ path })
    })
  }

  const component = renderer.create(ui)

  const snap = () => {
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  }

  snap()
  changeRoute('/bar')
  snap()
  changeRoute('/foo')
  snap()
})
