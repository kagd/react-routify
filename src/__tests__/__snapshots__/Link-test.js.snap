exports[`test Link should apply activeClassName 1`] = `
<a
  className="active"
  href="/nowhere"
  id="something" />
`;

exports[`test Link should merge className and activeClassName 1`] = `
<a
  className="foo active"
  href="/nowhere"
  id="something" />
`;

exports[`test Link should render 1`] = `
<a
  className="some-class"
  href="/nowhere"
  id="something" />
`;

exports[`test Link should update active state onRouteChange 1`] = `
<div>
  <a
    className="active"
    href="/foo">
    foo
  </a>
  <a
    href="/bar">
    bar
  </a>
</div>
`;

exports[`test Link should update active state onRouteChange 2`] = `
<div>
  <a
    href="/foo">
    foo
  </a>
  <a
    className="active"
    href="/bar">
    bar
  </a>
</div>
`;

exports[`test Link should update active state onRouteChange 3`] = `
<div>
  <a
    className="active"
    href="/foo">
    foo
  </a>
  <a
    href="/bar">
    bar
  </a>
</div>
`;
